### ISSUE TITLE
TLS exception

### ISSUE THREAD

**Issue by crazydef** [randombit/botan#3514]
When attempting to handshake a TLS 1.3 (or 1.2) connection, I am getting the following exception:

```
throw Invalid_Argument("to_u32bit invalid decimal string '" + str + "'");
```
from line 40 in utils/parsing.cpp.

The `str` in question is "DNS", coming from the following code:
```
bool X509_Certificate::matches_dns_name(std::string_view name) const
   {
   if(name.empty())
      return false;

   auto issued_names = subject_info("DNS");
   ...
```
where the incoming `name` is the server name I pass into the `Botan::TLS::Client` constructor through `Botan::TLS::Server_Information`.

**Comment by randombit** [comment]
Is it possible to get a backtrace from the point the exception is thrown? As I do not see any path where `subject_info` or anything it calls would try to convert anything into an integer.

**Comment by crazydef** [comment]
Of course.  :)

Here's the full callstack from the point Botan code is entered:

```
 	RaiseException()	Unknown
 	_CxxThrowException(void * pExceptionObject, const _s__ThrowInfo * pThrowInfo) Line 75	C++
>	Botan::to_u32bit(std::basic_string_view<char,std::char_traits<char>> str_view) Line 40	C++
 	Botan::`anonymous namespace'::parse_oid_str(std::basic_string_view<char,std::char_traits<char>> oid) Line 46	C++
 	Botan::OID::from_string(std::basic_string_view<char,std::char_traits<char>> str) Line 87	C++
 	Botan::X509_DN::has_field(std::basic_string_view<char,std::char_traits<char>> attr) Line 129	C++
 	Botan::X509_Certificate::subject_info(std::basic_string_view<char,std::char_traits<char>> req) Line 594	C++
 	Botan::X509_Certificate::matches_dns_name(std::basic_string_view<char,std::char_traits<char>> name) Line 676	C++
 	Botan::PKIX::check_chain(const std::vector<Botan::X509_Certificate,std::allocator<Botan::X509_Certificate>> & cert_path, std::chrono::time_point<std::chrono::system_clock,std::chrono::duration<__int64,std::ratio<1,10000000>>> ref_time, std::basic_string_view<char,std::char_traits<char>> hostname, Botan::Usage_Type usage, const Botan::Path_Validation_Restrictions & restrictions) Line 47	C++
 	Botan::x509_path_validate(const std::vector<Botan::X509_Certificate,std::allocator<Botan::X509_Certificate>> & end_certs, const Botan::Path_Validation_Restrictions & restrictions, const std::vector<Botan::Certificate_Store *,std::allocator<Botan::Certificate_Store *>> & trusted_roots, std::basic_string_view<char,std::char_traits<char>> hostname, Botan::Usage_Type usage, std::chrono::time_point<std::chrono::system_clock,std::chrono::duration<__int64,std::ratio<1,10000000>>> ref_time, std::chrono::duration<__int64,std::ratio<1,1000>> ocsp_timeout, const std::vector<std::optional<Botan::OCSP::Response>,std::allocator<std::optional<Botan::OCSP::Response>>> & ocsp_resp) Line 963	C++
 	Botan::TLS::Callbacks::tls_verify_cert_chain(const std::vector<Botan::X509_Certificate,std::allocator<Botan::X509_Certificate>> & cert_chain, const std::vector<std::optional<Botan::OCSP::Response>,std::allocator<std::optional<Botan::OCSP::Response>>> & ocsp_responses, const std::vector<Botan::Certificate_Store *,std::allocator<Botan::Certificate_Store *>> & trusted_roots, Botan::Usage_Type usage, std::basic_string_view<char,std::char_traits<char>> hostname, const Botan::TLS::Policy & policy) Line 88	C++
 	Botan::TLS::Certificate_13::verify(Botan::TLS::Callbacks & callbacks, const Botan::TLS::Policy & policy, Botan::Credentials_Manager & creds, std::basic_string_view<char,std::char_traits<char>> hostname, bool use_ocsp) Line 132	C++
 	Botan::TLS::Client_Impl_13::handle(const Botan::TLS::Certificate_13 & certificate_msg) Line 487	C++
 	`Botan::TLS::Client_Impl_13::process_handshake_msg'::`2'::<lambda_1>::operator()<std::reference_wrapper<Botan::TLS::Certificate_13>>(std::reference_wrapper<Botan::TLS::Certificate_13> msg) Line 93	C++
 	std::invoke<`Botan::TLS::Client_Impl_13::process_handshake_msg'::`2'::<lambda_1>,std::reference_wrapper<Botan::TLS::Certificate_13>>(Botan::TLS::Client_Impl_13::process_handshake_msg::__l2::<lambda_1> && _Obj, std::reference_wrapper<Botan::TLS::Certificate_13> && _Arg1) Line 1589	C++
 	std::_Variant_dispatcher<std::integer_sequence<unsigned __int64,5>>::_Dispatch2<void,`Botan::TLS::Client_Impl_13::process_handshake_msg'::`2'::<lambda_1>,std::variant<std::reference_wrapper<Botan::TLS::Server_Hello_13>,std::reference_wrapper<Botan::TLS::Server_Hello_12>,std::reference_wrapper<Botan::TLS::Hello_Retry_Request>,std::reference_wrapper<Botan::TLS::Encrypted_Extensions>,std::reference_wrapper<Botan::TLS::Certificate_13>,std::reference_wrapper<Botan::TLS::Certificate_Request_13>,std::reference_wrapper<Botan::TLS::Certificate_Verify_13>,std::reference_wrapper<Botan::TLS::Finished_13>>,0>(Botan::TLS::Client_Impl_13::process_handshake_msg::__l2::<lambda_1> && _Obj, std::variant<std::reference_wrapper<Botan::TLS::Server_Hello_13>,std::reference_wrapper<Botan::TLS::Server_Hello_12>,std::reference_wrapper<Botan::TLS::Hello_Retry_Request>,std::reference_wrapper<Botan::TLS::Encrypted_Extensions>,std::reference_wrapper<Botan::TLS::Certificate_13>,std::reference_wrapper<Botan::TLS::Certificate_Request_13>,std::reference_wrapper<Botan::TLS::Certificate_Verify_13>,std::reference_wrapper<Botan::TLS::Finished_13>> && <_Args_0>) Line 1460	C++
 	std::_Visit_strategy<2>::_Visit2<void,std::_Meta_list<std::integer_sequence<unsigned __int64,0>,std::integer_sequence<unsigned __int64,1>,std::integer_sequence<unsigned __int64,2>,std::integer_sequence<unsigned __int64,3>,std::integer_sequence<unsigned __int64,4>,std::integer_sequence<unsigned __int64,5>,std::integer_sequence<unsigned __int64,6>,std::integer_sequence<unsigned __int64,7>,std::integer_sequence<unsigned __int64,8>>,`Botan::TLS::Client_Impl_13::process_handshake_msg'::`2'::<lambda_1>,std::variant<std::reference_wrapper<Botan::TLS::Server_Hello_13>,std::reference_wrapper<Botan::TLS::Server_Hello_12>,std::reference_wrapper<Botan::TLS::Hello_Retry_Request>,std::reference_wrapper<Botan::TLS::Encrypted_Extensions>,std::reference_wrapper<Botan::TLS::Certificate_13>,std::reference_wrapper<Botan::TLS::Certificate_Request_13>,std::reference_wrapper<Botan::TLS::Certificate_Verify_13>,std::reference_wrapper<Botan::TLS::Finished_13>>>(unsigned __int64 _Idx, Botan::TLS::Client_Impl_13::process_handshake_msg::__l2::<lambda_1> && _Obj, std::variant<std::reference_wrapper<Botan::TLS::Server_Hello_13>,std::reference_wrapper<Botan::TLS::Server_Hello_12>,std::reference_wrapper<Botan::TLS::Hello_Retry_Request>,std::reference_wrapper<Botan::TLS::Encrypted_Extensions>,std::reference_wrapper<Botan::TLS::Certificate_13>,std::reference_wrapper<Botan::TLS::Certificate_Request_13>,std::reference_wrapper<Botan::TLS::Certificate_Verify_13>,std::reference_wrapper<Botan::TLS::Finished_13>> && <_Args_0>) Line 1536	C++
 	std::_Visit_impl<9,void,std::_Meta_list<std::integer_sequence<unsigned __int64,0>,std::integer_sequence<unsigned __int64,1>,std::integer_sequence<unsigned __int64,2>,std::integer_sequence<unsigned __int64,3>,std::integer_sequence<unsigned __int64,4>,std::integer_sequence<unsigned __int64,5>,std::integer_sequence<unsigned __int64,6>,std::integer_sequence<unsigned __int64,7>,std::integer_sequence<unsigned __int64,8>>,`Botan::TLS::Client_Impl_13::process_handshake_msg'::`2'::<lambda_1>,std::variant<std::reference_wrapper<Botan::TLS::Server_Hello_13>,std::reference_wrapper<Botan::TLS::Server_Hello_12>,std::reference_wrapper<Botan::TLS::Hello_Retry_Request>,std::reference_wrapper<Botan::TLS::Encrypted_Extensions>,std::reference_wrapper<Botan::TLS::Certificate_13>,std::reference_wrapper<Botan::TLS::Certificate_Request_13>,std::reference_wrapper<Botan::TLS::Certificate_Verify_13>,std::reference_wrapper<Botan::TLS::Finished_13>>>(Botan::TLS::Client_Impl_13::process_handshake_msg::__l2::<lambda_1> && _Obj, std::variant<std::reference_wrapper<Botan::TLS::Server_Hello_13>,std::reference_wrapper<Botan::TLS::Server_Hello_12>,std::reference_wrapper<Botan::TLS::Hello_Retry_Request>,std::reference_wrapper<Botan::TLS::Encrypted_Extensions>,std::reference_wrapper<Botan::TLS::Certificate_13>,std::reference_wrapper<Botan::TLS::Certificate_Request_13>,std::reference_wrapper<Botan::TLS::Certificate_Verify_13>,std::reference_wrapper<Botan::TLS::Finished_13>> && <_Args_0>) Line 1584	C++
 	std::visit<`Botan::TLS::Client_Impl_13::process_handshake_msg'::`2'::<lambda_1>,std::variant<std::reference_wrapper<Botan::TLS::Server_Hello_13>,std::reference_wrapper<Botan::TLS::Server_Hello_12>,std::reference_wrapper<Botan::TLS::Hello_Retry_Request>,std::reference_wrapper<Botan::TLS::Encrypted_Extensions>,std::reference_wrapper<Botan::TLS::Certificate_13>,std::reference_wrapper<Botan::TLS::Certificate_Request_13>,std::reference_wrapper<Botan::TLS::Certificate_Verify_13>,std::reference_wrapper<Botan::TLS::Finished_13>>,void>(Botan::TLS::Client_Impl_13::process_handshake_msg::__l2::<lambda_1> && _Obj, std::variant<std::reference_wrapper<Botan::TLS::Server_Hello_13>,std::reference_wrapper<Botan::TLS::Server_Hello_12>,std::reference_wrapper<Botan::TLS::Hello_Retry_Request>,std::reference_wrapper<Botan::TLS::Encrypted_Extensions>,std::reference_wrapper<Botan::TLS::Certificate_13>,std::reference_wrapper<Botan::TLS::Certificate_Request_13>,std::reference_wrapper<Botan::TLS::Certificate_Verify_13>,std::reference_wrapper<Botan::TLS::Finished_13>> && <_Args_0>) Line 1628	C++
 	Botan::TLS::Client_Impl_13::process_handshake_msg(std::variant<Botan::TLS::Client_Hello_13,Botan::TLS::Client_Hello_12,Botan::TLS::Server_Hello_13,Botan::TLS::Server_Hello_12,Botan::TLS::Hello_Retry_Request,Botan::TLS::Encrypted_Extensions,Botan::TLS::Certificate_13,Botan::TLS::Certificate_Request_13,Botan::TLS::Certificate_Verify_13,Botan::TLS::Finished_13> message) Line 83	C++
 	Botan::TLS::Channel_Impl_13::from_peer(std::span<unsigned char const ,-1> data) Line 139	C++
 	Botan::TLS::Client::from_peer(std::span<unsigned char const ,-1> data) Line 91	C++
 	Botan::TLS::Channel::received_data(const unsigned char * buf, unsigned __int64 buf_size) Line 50	C++
```

The call into `X509_DN::deref_info_field` from `X509_DN::has_field` returns the input string, "DNS", which `OID::from_string` then attempts to parse.

**Comment by randombit** [comment]
I can replicate this in that the exception is thrown, but it is immediately caught and another codepath taken. Is somehow this exception making its way up to your application?

**Comment by crazydef** [comment]
Ah, no.  My mistake.  I got an exception on first run, enabled exception catching in the debugger, and that was the first thing I saw.

My actual problem is this exception:
```
      throw TLS_Exception(Alert::BadCertificate, "Certificate validation failure: " + result.result_string());
```
where result contains OCSP_ISSUER_NOT_FOUND.

**Comment by randombit** [comment]
Ah OK well that makes sense at least. Can you share what site you are attempting to connect to? Also, which certificate stores are you enabling? (And maybe a more specific question on this, are you including a `System_Certificate_Store` in your list of trusted roots?)

**Comment by crazydef** [comment]
This is the page I'm attempting to fetch:  https://osocorporation.com/index.php

For the certificates, I literally just copied the Client_Credentials class from the tls_client example code.  (On the (possibly incorrect) assumption that that would be sufficient to get something up and running quickly.)  It only includes a System_Certificate_Store, so I thought (assumed) that would suffice.

**Comment by randombit** [comment]
That should be sufficient. But I'm also seeing cert issuer not found errors if I connect (on a Linux machine) using the `tls_client` cli. But Firefox and `openssl s_client` both seem to be happy. I'm not sure what's going on.

OK I just noticed that `tls_client google.com` also fails with issuer not found errors. WTF, something is broken here.

**Comment by randombit** [comment]
@reneme One probable bug here that I'm confused by - `Basic_Credentials_Manager::trusted_certificate_authorities` has logic to skip returning a certificate store if the `type` is `tls-server`. This has the intent of disabling requesting client auth. However in `tls_client`, we get invoked with `type == "tls-server"`! In `Certficate_13::verify`, does `m_side` represent *our* side or the *senders side*?

**Comment by randombit** [comment]
@reneme Another issue, I think related to the OCSP rework. `is_issued_by` will return false if the OCSP response doesn't include either the key hash or the issuer name. However that seems wrong in the case that the OCSP response is not signed by a delegated sub-CA.

**Comment by randombit** [comment]
Sorry about this. This should have been caught by testing before release, and just wasn't. Can you try the patch in #3515 and see if it helps?

**Comment by crazydef** [comment]
Yep, that seems to be working for me now.  :)

**Comment by crazydef** [comment]
No need to apologise, btw.  These things happen.  I'm just glad it wasn't something I'd screwed up.  :)

**Comment by reneme** [comment]
That's fixed with #3515, no? Please re-open if it isn't.

**Comment by crazydef** [comment]
Yes, it's fixed.
